pkg/operator/controller/machineconfig/nodes.go

```diff
@@ -25,109 +25,128 @@ import (
+
 	"sigs.k8s.io/controller-runtime/pkg/client"
-)
 
-// ensureReqNodeLabelExists is for checking the if the required labels exist on the nodes.
-// Returns true if all the targeted nodes have the required labels.
-// Return false if not all the nodes have the labels or none were selected.
-func (r *MachineConfigReconciler) ensureReqNodeLabelExists(ctx context.Context) (bool, error) {
+	"github.com/openshift/node-observability-operator/api/v1alpha2"
+)
 
-	// the number of the nodes which already had the required labels
-	// and the number of the nodes updated with the labels
-	existingNodeCount, updNodeCount := 0, 0
-	nodeList := &corev1.NodeList{}
-	if err := r.listNodes(ctx, nodeList, r.CtrlConfig.Spec.NodeSelector); err != nil {
-		return false, err
-	}
+const (
+	patchAdd    = "add"
+	patchRemove = "remove"
+)
 
-	if len(nodeList.Items) == 0 {
-		r.Log.V(1).Info("No nodes matching the given selector were found", "NodeSelector", r.CtrlConfig.Spec.NodeSelector)
-		return false, nil
-	}
+// ensureReqNodeLabelExists updates the given nodes with the required labels.
+func (r *MachineConfigReconciler) ensureReqNodeLabelExists(ctx context.Context, nodeList *corev1.NodeList) error {
 
+	nodeCount := 0
+	requiredNodeCount := len(nodeList.Items)
 	for i, node := range nodeList.Items {
 		if _, exist := node.Labels[NodeObservabilityNodeRoleLabelName]; exist {
-			existingNodeCount++
+			nodeCount++
 			continue
 		}
 
-		patch, _ := newPatch(add,
-			ResourceLabelsPath,
-			map[string]interface{}{
-				NodeObservabilityNodeRoleLabelName: Empty,
-			})
-		if err := r.ClientPatch(ctx, &nodeList.Items[i], client.RawPatch(types.JSONPatchType, patch)); err != nil {
-			return false, err
+		err := r.patchNodeLabels(ctx, &nodeList.Items[i], patchAdd, ResourceLabelsPath, map[string]interface{}{NodeObservabilityNodeRoleLabelName: empty})
+		if err != nil {
+			return fmt.Errorf("failed to patch node due to %w", err)
 		}
-		r.Log.V(1).Info("Successfully added label", "Node", node.Name, "Label", NodeObservabilityNodeRoleLabelName)
-
-		updNodeCount++
+		r.Log.V(1).Info("successfully performed add operation for labels", "Node", node.Name, "Label", NodeObservabilityNodeRoleLabelName)
+		nodeCount++
 	}
 
-	if (existingNodeCount == len(nodeList.Items)) || (updNodeCount == len(nodeList.Items)) {
-		r.Log.V(1).Info("Nodeobservability role is present on all the nodes with worker role", "ExistingNodeCount", existingNodeCount, "UpdatedNodeCount", updNodeCount)
-		return true, nil
+	if nodeCount == requiredNodeCount {
+		r.Log.V(1).Info("nodeobservability role is present on all the nodes with worker role", "UpdatedNodeCount", nodeCount)
+		return nil
 	}
 
-	return false, nil
+	r.Log.V(1).Info("nodeobservability role is not present on all the nodes with worker role", "Nodes", requiredNodeCount, "UpdatedNodeCount", nodeCount)
+	return fmt.Errorf("failed to ensure required label, expected updates on %d nodes but got %d", requiredNodeCount, nodeCount)
```

Comment:
I believe the error will be returned back to `controller-runtime` which means that it'll be logged, making the log before redundant.